### PR TITLE
Product Block Editor: introduce AttributesComboboxControl component

### LINKED ISSUE #46913: Product Block Editor: introduce component based on ComboboxControl core component

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces the new `<AttributesComboboxControl />` components to be used in the admin area of the Product Attributes of the application (follow-up).
Its implementation is based on the ComboboxControl core component.

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Test using storybook.

1. Build the Woocommerce plugin

```cli
pnpm --filter='@woocommerce/plugin-woocommerce' build
```

2. Run Storybook

```cli
pnpm --filter='@woocommerce/storybook' storybook
```

4. Go to the AttributesComboboxControl stories
```
http://localhost:6007/?path=/story/product-editor-components-attributescomboboxcontrol--default
```
5. Confirm how the component works

image

_**Disclaimer: The storybook context differs from the product editor's. Styles, components, etc., behave differently, affecting the UI/UX of it. Anyway, the story is good enough to test part of the implementation.**_ 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/components/attribute-combobox-field/types.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import type { ProductAttribute } from '@woocommerce/data';
+
+/*
+ * Define the attributes combobox control item type,
+ * which is a combination of the product attribute and
+ * additional properties.
+ */
+export type AttributesComboboxControlItem = ProductAttribute & {
+	isDisabled?: boolean;
+	takenBy?: number;
```
Comment (nathanss) [inline packages/js/product-editor/src/components/attribute-combobox-field/types.ts]: What is `takenBy`? It's not yet used, right? I couldn't figure out just by the name.

File: packages/js/product-editor/src/components/attribute-combobox-field/styles.scss [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	.components-form-token-field__suggestion {
+		padding: 0;
+		min-height: 32px;
+		display: flex;
+		align-items: center;
+
+		.item-wrapper {
+			padding: 9px 12px;
```
Comment (nathanss) [inline packages/js/product-editor/src/components/attribute-combobox-field/styles.scss]: Any reason why this is 9? Can we change to 8px?

File: packages/js/product-editor/src/components/attribute-combobox-field/styles.scss [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		height: 36px;
+	}
+
+	.components-combobox-control__suggestions-container {
+		position: absolute;
+		padding: 0;
+		background-color: white;
+
+		> .components-flex {
+			height: 34px;
```
Comment (nathanss) [inline packages/js/product-editor/src/components/attribute-combobox-field/styles.scss]: Can we change this to 32px?

**Review by nathanss — DISMISSED** [review]
Overall, it looks very promising! Looking forward to see this developed further and used in practice.

~~I found a small UI issue, there is a grey margin between the input (white) and the border. Is it supposed to look like this?~~
EDIT: just read your disclaimer. If it looks fine in the product editor, feel free to ignore the comment, sorry about that :)

Also left some small suggestions

File: packages/js/product-editor/src/components/attribute-combobox-field/index.tsx [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+				label:
+					createNewAttributeOption.state === 'draft'
+						? sprintf(
+								/* translators: The name of the new attribute term to be created */
+								__( 'Create "%s"', 'woocommerce' ),
+								createNewAttributeOption.label
+						  )
+						: createNewAttributeOption.label,
+				value: createNewAttributeOption.value,
+			},
```
Comment (louwie17) [inline packages/js/product-editor/src/components/attribute-combobox-field/index.tsx:145]: Adding this should depend on if the existing item doesn't exist.
Maybe this should be an option? `allowDuplicates`?

File: packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	);
+};
+
+Default.args = {
+	label: __( 'Attributes', 'woocommerce' ),
+	items,
+	help: __( 'Select or create attributes for this product.', 'woocommerce' ),
+
+	onChange: ( newValue: ProductAttribute ) => {
+		console.log( '(onChange) newValue:', newValue ); // eslint-disable-line no-console
```
Comment (louwie17) [inline packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx:111]: I thought we could make use of actions here: https://storybook.js.org/docs/essentials/actions

**Review by louwie17 — APPROVED** [review]
Approved as this is just in the story book, but did suggest two follow ups within the code.

File: packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx (hunk 5 above)
Comment (retrofox) [reply packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx:111]: I tried in the past, but it didn't work. Let me try it again 🤞

File: packages/js/product-editor/src/components/attribute-combobox-field/index.tsx (hunk 4 above)
Comment (retrofox) [reply packages/js/product-editor/src/components/attribute-combobox-field/index.tsx:145]: Do you mean allowing the creation of a new attribute with the same `name`? 

This is something that the current implementation allows by default. However.

image

Above, you can create attributes with the same name by clicking on the `Create "..."` again and again.

 I've just realized that the attributes wp-admin page doesn't when the slug is not explicitly defined:

image

`allowDuplicates` or something more specific to the attribute name. we should create a new issue and add it to the tasks list https://github.com/woocommerce/woocommerce/issues/45294

File: packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx (hunk 5 above)
Comment (retrofox) [reply packages/js/product-editor/src/components/attribute-combobox-field/stories/index.tsx:111]:
```
import { fn } from '@storybook/test';
```

The app doesn't have defined the `@storybook/test`dependency. We can do it in a follow up.

File: packages/js/product-editor/src/components/attribute-combobox-field/index.tsx (hunk 4 above)
Comment (louwie17) [reply packages/js/product-editor/src/components/attribute-combobox-field/index.tsx:145]: Oooh yeah, I think we do allow this in variation options but not in attributes.
So we did need a way to toggle that.

Comment (retrofox) [reply packages/js/product-editor/src/components/attribute-combobox-field/index.tsx:145]: thanks, it makes totally sense

Comment (retrofox) [reply packages/js/product-editor/src/components/attribute-combobox-field/index.tsx:145]: https://github.com/woocommerce/woocommerce/issues/46954

**Pull request merged**